### PR TITLE
test: BCowPool integration tests

### PR DESCRIPTION
**Pull request by 0xAustrian**
Closes BAL-88

### CODE REVIEW

File: test/integration/BCowPool.t.sol [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    // deal LP
+    deal(address(dai), address(lp), 100e18);
+    deal(address(weth), address(lp), 100e18);
+
+    // deal swapper
+    deal(address(weth), address(swapper), 1e18);
+
+    vm.startPrank(controller);
+    // deploy
+    pool = new BCoWPool(address(settlement));
```
Comment (wei3erHase) [inline test/integration/BCowPool.t.sol:45]: blz add a `TODO: deploy with BCoWFactory` comment

File: test/integration/BCowPool.t.sol [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    // enable trading
+    pool.enableTrading(APP_DATA_HASH);
```
Comment (wei3erHase) [inline test/integration/BCowPool.t.sol:47]: hmm, interesting, this is not supposed to happen, perhaps we could `enableTrading` only after `finalize` was called

File: test/integration/BCowPool.t.sol [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    vm.prank(swapper);
+    // TODO: add vaultRelayer() to BCoWPool interface
+    // weth.approve(pool.vaultRelayer(), type(uint256).max);
+
+    // swapper creates the order
+    // TODO: add MAX_ORDER_DURATION() to BCoWPool interface
+    uint32 latestValidTimestamp = uint32(block.timestamp); // + pool.MAX_ORDER_DURATION();
+    GPv2Order.Data memory order = GPv2Order.Data({
+      sellToken: weth,
+      buyToken: dai,
```
Comment (wei3erHase) [inline test/integration/BCowPool.t.sol:90]: sell 100 weth for 1 dai?!

File: test/integration/BCowPool.t.sol [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      receiver: GPv2Order.RECEIVER_SAME_AS_OWNER,
+      sellAmount: sellAmount,
+      buyAmount: buyAmount,
+      validTo: latestValidTimestamp,
+      appData: APP_DATA_HASH,
```
Comment (wei3erHase) [inline test/integration/BCowPool.t.sol:95]: i believe this is supossed to be `APP_DATA` (not the hash)

File: test/integration/BCowPool.t.sol (hunk 2 above)
Comment (0xAustrian) [reply test/integration/BCowPool.t.sol:47]: Will move it after finalize and we can add a revert inside `enableTrading` later

File: test/integration/BCowPool.t.sol (hunk 3 above)
Comment (0xAustrian) [reply test/integration/BCowPool.t.sol:90]: What? did you not hear the news??

File: test/integration/BCowPool.t.sol [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+      target: address(pool),
+      value: 0,
+      callData: abi.encodeWithSelector(
+        IBCoWPool.commit.selector, poolOrder.hash(pool.SOLUTION_SETTLER_DOMAIN_SEPARATOR())
+      )
+    });
+
+    // finally, settle
+    vm.startPrank(solver);
+    settlement.settle(tokens, clearingPrices, trades, interactions);
```
Comment (wei3erHase) [inline test/integration/BCowPool.t.sol:172]: add gas snapshot! 🙏

File: test/integration/BCowPool.t.sol [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    });
+    bytes memory poolSig = abi.encode(poolOrder);
+
+    // solver prepares for call settle()
+    IERC20[] memory tokens = new IERC20[](2);
+    tokens[0] = IERC20(weth);
+    tokens[1] = IERC20(dai);
+
+    uint256[] memory clearingPrices = new uint256[](2);
+    // TODO: we can use more accurate clearing prices here
```
Comment (wei3erHase) [inline test/integration/BCowPool.t.sol:122]: is ok, i think ideally we could merge this test with `PoolSwap.t.sol`, in a "regression" way, such that we have the same expectations for swaps, but let's make it on another PR (and perhaps after BCoWFactory) is done

File: test/integration/BCowPool.t.sol [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    GPv2Interaction.Data[][3] memory interactions =
+      [new GPv2Interaction.Data[](1), new GPv2Interaction.Data[](0), new GPv2Interaction.Data[](0)];
+
+    interactions[0][0] = GPv2Interaction.Data({
+      target: address(pool),
+      value: 0,
+      callData: abi.encodeWithSelector(
+        IBCoWPool.commit.selector, poolOrder.hash(pool.SOLUTION_SETTLER_DOMAIN_SEPARATOR())
+      )
+    });
```
Comment (wei3erHase) [inline test/integration/BCowPool.t.sol:168]: hmmmmm interesting, i just worry about: what avoids the settler from adding
```solidity
  interactions[0][0] = GPv2Interaction.Data({
    target: address(DAI),
    value: 0,
    callData: abi.encodeWithSelector(
      ERC20.transferFrom.selector, address(pool), address(bunny), 100000
    )
```

File: test/integration/GPv2TradeEncoder.sol
[the remark this answers is not in the document]
Comment (wei3erHase) [reply test/integration/GPv2TradeEncoder.sol:1]: ideally we would vendor it, they told me to try sth like this on remappings, but am unsure about it:
```
  "lib/composable-cow:cowprotocol/=lib/composable-cow/lib/cowprotocol/src/contracts/",
  "lib/composable-cow:safe/=lib/composable-cow/lib/safe/contracts/",
  ```

give it a try, else we're fine

File: test/integration/BCowPool.t.sol (hunk 7 above)
Comment (wei3erHase) [reply test/integration/BCowPool.t.sol:168]: ahá! the `solutionSettler` contract doesn't have the allowances :) the `vaultRelayer` has

**Review by wei3erHase — APPROVED** [review]
approved gud ser, greatly done! 👏
